In [1]:
import random

import numpy as np
import pandas as pd

from dataset import get_dataset
from trainer import Trainer
from loss import MyLoss
from model import MyModel

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import models

In [2]:
## seeds

random_seed = 42

torch.manual_seed(random_seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
np.random.seed(random_seed)
random.seed(random_seed)

In [3]:
## hyperparameters

batch_size = 64
lr = 0.00006
num_epoch = 40

df_train_path = '/opt/ml/code/df/df_train_cluster.csv'
df_valid_path = '/opt/ml/code/df/df_valid_cluster.csv'

In [4]:
## prepare dataset

df_train = pd.read_csv(df_train_path)
df_valid = pd.read_csv(df_valid_path)

train_dataset, valid_dataset = get_dataset(df_train, df_valid)

In [5]:
## prepare Dataloader

dataloaders = {'train' : DataLoader(train_dataset, batch_size=batch_size, num_workers=3, drop_last=True),
               'valid' : DataLoader(valid_dataset, batch_size=batch_size, shuffle=False, num_workers=3)}

In [6]:
## prepare model

device = 'cuda'

model = MyModel('efficientnet-b0', 100).to(device)
print('ready')

Loaded pretrained weights for efficientnet-b0
ready


In [7]:
# check input train batch

# from utils import imshow
# i1, i2, l = next(iter(dataloaders['train']))
# print(l)
# imshow(i2)

In [8]:
criterion = MyLoss(smoothing=0.1)
optimizer = optim.Adam(model.parameters(), lr=lr)
lr_scheduler = None #optim.lr_scheduler.MultiStepLR(optimizer, milestones=[5, 10, 15], gamma = 0.2)

## Train

In [9]:
trainer = Trainer('/opt/ml/code/save')

In [14]:
## resume
# model.load_state_dict(torch.load('/opt/ml/code/save/label/effnet_test018.pt'))

# change lr manually:
for g in optimizer.param_groups:
    g['lr'] = 0.00001

In [15]:
trainer.train(model, dataloaders, df_valid, criterion, optimizer, device, num_epochs=num_epoch, scheduler=lr_scheduler,
              sub_dir='cluster', save_name='epoch_')

Epoch 1/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.76it/s]


[Train]  Loss: 0.0243 D1: 0.0843 D0: 0.9485


100%|██████████| 60/60 [00:08<00:00,  7.50it/s]


[Valid]  Acc: 0.9195 Margin: 0.9687

Epoch 2/40
----------


100%|██████████| 124/124 [01:11<00:00,  1.74it/s]


[Train]  Loss: 0.0221 D1: 0.0812 D0: 0.9519


100%|██████████| 60/60 [00:08<00:00,  7.50it/s]


[Valid]  Acc: 0.9195 Margin: 0.9769

Epoch 3/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.79it/s]


[Train]  Loss: 0.0222 D1: 0.0795 D0: 0.9508


100%|██████████| 60/60 [00:08<00:00,  7.45it/s]


[Valid]  Acc: 0.9192 Margin: 0.9810

Epoch 4/40
----------


100%|██████████| 124/124 [01:11<00:00,  1.74it/s]


[Train]  Loss: 0.0235 D1: 0.0827 D0: 0.9492


100%|██████████| 60/60 [00:08<00:00,  7.45it/s]


[Valid]  Acc: 0.9207 Margin: 0.9911

Epoch 5/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.77it/s]


[Train]  Loss: 0.0222 D1: 0.0806 D0: 0.9511


100%|██████████| 60/60 [00:08<00:00,  7.48it/s]


[Valid]  Acc: 0.9207 Margin: 0.9764

Epoch 6/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0230 D1: 0.0816 D0: 0.9502


100%|██████████| 60/60 [00:08<00:00,  7.46it/s]


[Valid]  Acc: 0.9202 Margin: 0.9500

Epoch 7/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0227 D1: 0.0794 D0: 0.9492


100%|██████████| 60/60 [00:07<00:00,  7.53it/s]


[Valid]  Acc: 0.9190 Margin: 0.9708

Epoch 8/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0215 D1: 0.0801 D0: 0.9545


100%|██████████| 60/60 [00:08<00:00,  7.46it/s]


[Valid]  Acc: 0.9200 Margin: 0.9685

Epoch 9/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.79it/s]


[Train]  Loss: 0.0206 D1: 0.0787 D0: 0.9528


100%|██████████| 60/60 [00:08<00:00,  7.48it/s]


[Valid]  Acc: 0.9205 Margin: 0.9691

Epoch 10/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.76it/s]


[Train]  Loss: 0.0219 D1: 0.0811 D0: 0.9516


100%|██████████| 60/60 [00:07<00:00,  7.51it/s]


[Valid]  Acc: 0.9202 Margin: 0.9545

Epoch 11/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0203 D1: 0.0764 D0: 0.9532


100%|██████████| 60/60 [00:08<00:00,  7.48it/s]


[Valid]  Acc: 0.9205 Margin: 0.9566

Epoch 12/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0221 D1: 0.0795 D0: 0.9509


100%|██████████| 60/60 [00:07<00:00,  7.51it/s]


[Valid]  Acc: 0.9202 Margin: 0.9587

Epoch 13/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0208 D1: 0.0796 D0: 0.9541


100%|██████████| 60/60 [00:08<00:00,  7.44it/s]


[Valid]  Acc: 0.9203 Margin: 0.9584

Epoch 14/40
----------


100%|██████████| 124/124 [01:11<00:00,  1.74it/s]


[Train]  Loss: 0.0221 D1: 0.0768 D0: 0.9497


100%|██████████| 60/60 [00:07<00:00,  7.53it/s]


[Valid]  Acc: 0.9208 Margin: 0.9661

Epoch 15/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0219 D1: 0.0804 D0: 0.9523


100%|██████████| 60/60 [00:08<00:00,  7.46it/s]


[Valid]  Acc: 0.9220 Margin: 0.9638

Epoch 16/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0213 D1: 0.0792 D0: 0.9532


100%|██████████| 60/60 [00:08<00:00,  7.49it/s]


[Valid]  Acc: 0.9210 Margin: 0.9654

Epoch 17/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0224 D1: 0.0783 D0: 0.9501


100%|██████████| 60/60 [00:08<00:00,  7.43it/s]


[Valid]  Acc: 0.9219 Margin: 0.9794

Epoch 18/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0228 D1: 0.0829 D0: 0.9526


100%|██████████| 60/60 [00:08<00:00,  7.47it/s]


[Valid]  Acc: 0.9230 Margin: 0.9793

Epoch 19/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0222 D1: 0.0808 D0: 0.9518


100%|██████████| 60/60 [00:08<00:00,  7.45it/s]


[Valid]  Acc: 0.9229 Margin: 0.9715

Epoch 20/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0217 D1: 0.0800 D0: 0.9529


100%|██████████| 60/60 [00:08<00:00,  7.50it/s]


[Valid]  Acc: 0.9228 Margin: 0.9771

Epoch 21/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0206 D1: 0.0790 D0: 0.9546


100%|██████████| 60/60 [00:08<00:00,  7.46it/s]


[Valid]  Acc: 0.9231 Margin: 0.9693

Epoch 22/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0211 D1: 0.0784 D0: 0.9520


100%|██████████| 60/60 [00:08<00:00,  7.45it/s]


[Valid]  Acc: 0.9227 Margin: 0.9858

Epoch 23/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0215 D1: 0.0803 D0: 0.9548


100%|██████████| 60/60 [00:08<00:00,  7.46it/s]


[Valid]  Acc: 0.9223 Margin: 0.9842

Epoch 24/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0224 D1: 0.0787 D0: 0.9509


100%|██████████| 60/60 [00:07<00:00,  7.51it/s]


[Valid]  Acc: 0.9236 Margin: 0.9790

Epoch 25/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.79it/s]


[Train]  Loss: 0.0199 D1: 0.0784 D0: 0.9550


100%|██████████| 60/60 [00:08<00:00,  7.45it/s]


[Valid]  Acc: 0.9237 Margin: 0.9779

Epoch 26/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.79it/s]


[Train]  Loss: 0.0214 D1: 0.0800 D0: 0.9537


100%|██████████| 60/60 [00:07<00:00,  7.52it/s]


[Valid]  Acc: 0.9240 Margin: 0.9774

Epoch 27/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.76it/s]


[Train]  Loss: 0.0214 D1: 0.0771 D0: 0.9515


100%|██████████| 60/60 [00:07<00:00,  7.50it/s]


[Valid]  Acc: 0.9230 Margin: 0.9708

Epoch 28/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0207 D1: 0.0769 D0: 0.9524


100%|██████████| 60/60 [00:08<00:00,  7.49it/s]


[Valid]  Acc: 0.9231 Margin: 0.9765

Epoch 29/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0222 D1: 0.0785 D0: 0.9498


100%|██████████| 60/60 [00:08<00:00,  7.48it/s]


[Valid]  Acc: 0.9239 Margin: 0.9807

Epoch 30/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0214 D1: 0.0801 D0: 0.9544


100%|██████████| 60/60 [00:07<00:00,  7.52it/s]


[Valid]  Acc: 0.9232 Margin: 0.9813

Epoch 31/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0218 D1: 0.0786 D0: 0.9530


100%|██████████| 60/60 [00:08<00:00,  7.47it/s]


[Valid]  Acc: 0.9239 Margin: 0.9670

Epoch 32/40
----------


100%|██████████| 124/124 [01:08<00:00,  1.80it/s]


[Train]  Loss: 0.0214 D1: 0.0808 D0: 0.9539


100%|██████████| 60/60 [00:07<00:00,  7.58it/s]


[Valid]  Acc: 0.9243 Margin: 0.9688

Epoch 33/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0235 D1: 0.0810 D0: 0.9503


100%|██████████| 60/60 [00:07<00:00,  7.52it/s]


[Valid]  Acc: 0.9248 Margin: 0.9720

Epoch 34/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0205 D1: 0.0780 D0: 0.9549


100%|██████████| 60/60 [00:08<00:00,  7.49it/s]


[Valid]  Acc: 0.9242 Margin: 0.9710

Epoch 35/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.75it/s]


[Train]  Loss: 0.0217 D1: 0.0767 D0: 0.9510


100%|██████████| 60/60 [00:08<00:00,  7.46it/s]


[Valid]  Acc: 0.9241 Margin: 0.9745

Epoch 36/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0215 D1: 0.0789 D0: 0.9533


100%|██████████| 60/60 [00:07<00:00,  7.52it/s]


[Valid]  Acc: 0.9237 Margin: 0.9841

Epoch 37/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.76it/s]


[Train]  Loss: 0.0215 D1: 0.0787 D0: 0.9523


100%|██████████| 60/60 [00:07<00:00,  7.50it/s]


[Valid]  Acc: 0.9244 Margin: 0.9912

Epoch 38/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0212 D1: 0.0797 D0: 0.9545


100%|██████████| 60/60 [00:07<00:00,  7.50it/s]


[Valid]  Acc: 0.9243 Margin: 0.9782

Epoch 39/40
----------


100%|██████████| 124/124 [01:09<00:00,  1.78it/s]


[Train]  Loss: 0.0225 D1: 0.0792 D0: 0.9492


100%|██████████| 60/60 [00:08<00:00,  7.45it/s]


[Valid]  Acc: 0.9245 Margin: 0.9729

Epoch 40/40
----------


100%|██████████| 124/124 [01:10<00:00,  1.77it/s]


[Train]  Loss: 0.0193 D1: 0.0773 D0: 0.9567


100%|██████████| 60/60 [00:08<00:00,  7.42it/s]


[Valid]  Acc: 0.9235 Margin: 0.9721

Training complete!


## Cluster Experiment Logs

---

1. backbone =eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=100, bounding=clamp
    * epoch 8)
      - valid Loss: 0.1103 d1: 0.0020 d0: 0.8484544249352944
        
2. backbone =eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=100, bounding=sigmoid
    * epoch 32)
      - train Loss: 0.3030 d1: 0.2789 d0: 0.47928544043563304
      - valid Loss: 0.0707 d1: 0.0226 d0: 0.908650320648902
      
3. backbone =eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=100, bounding=tanh
    * epoch 45)
      - Train Loss: 0.0771 d1: 0.0895 d0: 0.9352
      - Valid acc : 0.8125 margin: 0.3051

4. backbone =eff b0, optimizer=Adam, lr=0.00006 -> 0.00001, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=200, bounding=tanh
    * epoch 35)
      - Train Loss: Loss: 0.0804 d1: 0.0959 d0: 0.9350
      - Valid acc : acc: 0.7941 margin: 0.3799
      
5. backbone=eff b1, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=200, bounding=tanh
    * 별로인듯
    
6. backbone =eff b0, optimizer=Adam, lr=0.0001 -> 0.00002, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=50, bounding=tanh
    * epoch 52)
      - Train Loss: 0.0665 D1: 0.0754 D0: 0.9424
      - Valid Acc: 0.8093 Margin: 0.2362
     
7. backbone =eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0), split 20%, hdim=100, bounding=tanh, vlaid_no_sigmoid
    * epoch 66)
      - [Train]  Loss: 0.0678 D1: 0.0786 D0: 0.9431
      -  valid : Acc: 0.8418 Margin: 0.5782

8. backbone =eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%, hdim=100, bounding=tanh, vlaid_no_sigmoid
    * epoch 92)
      - [Train]  Loss: 0.0349 D1: 0.0724 D0: 0.9507
      - [Valid]  Acc: 0.8831 Margin: 0.7154

9. backbone =eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.1), split 20%, hdim=100, bounding=tanh, vlaid_no_sigmoid
    * epoch 113)
      - [Train]  Loss: 0.0241 D1: 0.0832 D0: 0.9489
      - [Valid]  Acc: 0.9207 Margin: 0.9631
    * lr ---> 0.00001
      - [Valid]  Acc: 0.9248 Margin: 0.9720
      

## Debug

In [ ]:
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import confusion_matrix

model.eval()
y_true = []
y_pred = []
wrong = []
for inputs, labels in dataloaders['valid']:
    inputs = inputs.to(device)
    labels = labels.to(device)

    with torch.no_grad():
        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        y_true.extend(labels.tolist())
        y_pred.extend(preds.tolist())
        
        for p, t, inp in zip(preds, labels, inputs):
            if p.item() != t.item():
                wrong.append((inp.cpu().numpy(), p.item(), t.item()))

mtx = confusion_matrix(y_true, y_pred)
print(mtx)


In [ ]:
import matplotlib.pyplot as plt

label_ = list(range(6))

fig, ax = plt.subplots(figsize=(8,8))
im = ax.imshow(mtx, cmap='Blues', vmax=100, vmin=-20)

ax.set_xticks(np.arange(len(label_)))
ax.set_yticks(np.arange(len(label_)))
# ... and label them with the respective list entries
ax.set_xticklabels(label_)
ax.set_yticklabels(label_)

# Rotate the tick labels and set their alignment.
plt.setp(ax.get_xticklabels(), rotation=45, ha="right",
         rotation_mode="anchor")

# Loop over data dimensions and create text annotations.
for i in range(len(label_)):
    for j in range(len(label_)):
        if mtx[i, j]:
            text = ax.text(j, i, mtx[i, j],
                           ha="center", va="center", color="w", size=10)
ax.set_xlabel('Prediction')
ax.set_ylabel('True Label')
ax.set_title("Confusion mtx")
fig.tight_layout()
plt.show()

In [ ]:
import matplotlib.pyplot as plt
iterator = iter(wrong)

label = ["Male with mask", "Female with Mask", "Male incorrect mask", "Female no mask", "Male incorrect mask", "Female no mask"]
mean = np.array([0.485, 0.456, 0.406])
std = np.array([0.229, 0.224, 0.225])

In [ ]:
fig, axes = plt.subplots(2, 2, figsize=(10,10))
for ax in axes.flatten():
    img, p, t = next(iterator)
    img = img.transpose((1, 2, 0))
    img = std * img + mean
    img = np.clip(img, 0, 1)
    ax.imshow(img)
    ax.axis('off')
    ax.set_title(f't : {label[t]}\np : {label[p]}')
plt.show()

## Inference

In [ ]:
def sum_label(df):
    
    def _label(row):
        return row['age'] + 3*row['label']# + 6*row['mask']

    df['ans'] = df.apply(_label, axis=1)
    return df[['ImageID', 'ans']]

In [ ]:
## inference

# model.load_state_dict(torch.load('/opt/ml/code/save/best_log2.pt'))
# df_submit = create_label(model, test_dataloader, df_test.copy(), device, target='label')

# model = EfficientNet.from_pretrained(model_name, num_classes=3).to(device)
# model.load_state_dict(torch.load('/opt/ml/code/save/age/extended_far_best.pt'))
# df_submit = create_label(model, test_dataloader, df_submit, device, target='age')

# model.load_state_dict(torch.load('/opt/ml/code/save/mask/6e5_9987.pt'))
# df_submit = create_label(model, test_dataloader, df_submit, device, target='mask')

# model.fc = nn.Linear(model.fc.in_features, 2).cuda()
# model.load_state_dict(torch.load('/opt/ml/code/save/gender/6e5_9851.pt'))
# df_submit = create_label(model, test_dataloader, df_submit, device, target='gender')


# df_submit = sum_label(df_submit)
# df_submit


df_submit.to_csv('submit.csv', index=False)